src/serialization/gpsw_ser.c

```diff
@@ -0,0 +1,258 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <cifer/abe/gpsw.h>
+#include "cifer/internal/common.h"
+#include <cifer/serialization/gpsw_ser.h>
+#include <cifer/serialization/gpsw.pb-c.h>
+
+void cfe_gpsw_pub_key_pack(cfe_gpsw_pub_key *a, CfeGpswPubKeySer *msg, OctetSer *octets_val) {
+    cfe_FP12_BN254_pack(&(a->y), msg->y);
+
+    cfe_vec_octet octets;
+    cfe_vec_octet_init(&octets, a->t.size);
+    for (size_t i =0; i<a->t.size; i++) {
+        char *h1 = (char *) cfe_malloc((4 * MODBYTES_256_56) * sizeof(char));
+        octet oct = {0, (4 * MODBYTES_256_56) * sizeof(char), h1};
+        ECP2_BN254_toOctet(&oct, &(a->t.vec[i]));
+        oct.max = oct.len;
+        octets.vec[i] = oct;
+    }
+    cfe_vec_octet_pack(&octets, msg->t, octets_val);
+
+    for (size_t i =0; i<a->t.size; i++) {
+        free(octets.vec[i].val);
+    }
+    free(octets.vec);
+}
+
+void cfe_gpsw_pub_key_ser(cfe_gpsw_pub_key *a, cfe_ser *buf) {
+    CfeGpswPubKeySer msg = CFE_GPSW_PUB_KEY_SER__INIT;
+    VecOctetSer t = VEC_OCTET_SER__INIT;
+    msg.t = &t;
+    OctetSer y = OCTET_SER__INIT;
+    msg.y = &y;
+    OctetSer *octets_val = cfe_malloc(sizeof(OctetSer) * a->t.size);
+
+    cfe_gpsw_pub_key_pack(a, &msg, octets_val);
+
+    buf->len = cfe_gpsw_pub_key_ser__get_packed_size(&msg);
+    buf->ser = cfe_malloc(buf->len);
+
+    cfe_gpsw_pub_key_ser__pack(&msg, buf->ser);
+
+    for (size_t i =0; i<a->t.size; i++) {
+        free(msg.t->vec[i]->val);
+    }
+    free(msg.t->vec);
+    free(msg.y->val);
+    free(octets_val);
+}
+
+void cfe_gpsw_pub_key_unpack(cfe_gpsw_pub_key *a, CfeGpswPubKeySer *msg) {
+    cfe_FP12_BN254_unpack(&(a->y), msg->y);
+
+    cfe_vec_octet octets;
+    cfe_vec_octet_unpack(&octets, msg->t);
+    cfe_vec_G2_init(&(a->t), octets.size);
+    for (size_t i =0; i< octets.size; i++) {
+        ECP2_BN254_fromOctet(&(a->t.vec[i]), &(octets.vec[i]));
+        free(octets.vec[i].val);
+    }
+    cfe_vec_octet_free(&octets);
+}
+
+cfe_error cfe_gpsw_pub_key_read(cfe_gpsw_pub_key *a, cfe_ser *buf) {
+    CfeGpswPubKeySer *msg;
+    msg = cfe_gpsw_pub_key_ser__unpack(NULL, buf->len, buf->ser);
+    if (msg == NULL) {
+        return 1;
+    }
+
+    cfe_gpsw_pub_key_unpack(a, msg);
+    // Free the unpacked message
+    cfe_gpsw_pub_key_ser__free_unpacked(msg, NULL);
+    return 0;
+}
+
+void cfe_gpsw_key_pack(cfe_gpsw_key *a, CfeGpswKeySer *msg, MpzSer *val, OctetSer *octets_val) {
+    cfe_msp_pack(&(a->msp), msg->msp, val);
```

Comment:
You are packing `msp` again here. So just remove this line (and then probably you do not need `val` in the args) and it should be fine.

Reply:
Thanks! It worked like a charm :wink: